client/src/components/UserContext.tsx

```diff
@@ -0,0 +1,46 @@
+import { ReactNode, createContext, useCallback, useState } from 'react';
+import { removeAuth, saveAuth } from '../lib';
+
+export type User = {
+  userId: number;
+  username: string;
+  displayName: string;
+};
+
+export type UserContextValues = {
+  user: User | undefined;
+  token: string | undefined;
+  handleSignIn: (user: User, token: string) => void;
+  handleSignOut: () => void;
+};
+export const UserContext = createContext<UserContextValues>({
+  user: undefined,
+  token: undefined,
+  handleSignIn: () => undefined,
+  handleSignOut: () => undefined,
+});
+
+type Props = {
+  children: ReactNode;
+};
+export function UserProvider({ children }: Props) {
+  const [user, setUser] = useState<User>();
+  const [token, setToken] = useState<string>();
+
+  const handleSignIn = useCallback((user: User, token: string) => {
+    setUser(user);
+    setToken(token);
+    saveAuth(user, token);
+  }, []);
+
+  function handleSignOut() {
```

Comment:
It would be a good idea to put this one in a `useCallback`, also.

Technically, the `handleSignIn` and `handleSignOut` functions only need to be in `useCallback` if they ever get put into a dependency array (such as in a `useEffect`). It's unlikely these will be. But it's safe to always wrap your context function values in `useCallback`, just in case.

Reply:
I did actually put handleSignIn in a useEffect for the local storage login, which is why I changed it to a useCallback.